src/dbnode/persist/fs/commitlog/commit_log.go

```diff
@@ -506,69 +541,169 @@ func (l *commitLog) write() {
-
-			// If we failed to close then create a new commit log writer
-			l.writerState.writer = nil
+func (l *commitLog) openWriters() (persist.CommitLogFiles, error) {
+	// Ensure that the previous asynchronous reset of the secondary writer (if any)
+	// has completed before attempting to start a new one and/or modify the writerState
+	// in any way.
+	l.waitForSecondaryWriterAsyncResetComplete()
+
+	if l.writerState.primaryWriter.writer == nil || l.writerState.secondaryWriter.writer == nil {
+		if l.writerState.primaryWriter.writer != nil {
+			// Make sure to close and flush any remaining data before creating a new writer if the
+			// primary (which contains data) is not nil.
+			if err := l.writerState.primaryWriter.writer.Close(); err != nil {
+				return nil, err
+			}
 		}
-	}
 
-	if l.writerState.writer == nil {
-		l.writerState.writer = l.newCommitLogWriterFn(l.onFlush, l.opts)
-	}
+		if l.writerState.secondaryWriter.writer != nil {
+			// Ignore errors because the secondary file doesn't have any data.
+			l.writerState.secondaryWriter.writer.Close()
+		}
 
-	file, err := l.writerState.writer.Open()
-	if err != nil {
-		return persist.CommitLogFile{}, err
+		// If either of the commitlog writers is nil then open both of them synchronously. Under
+		// normal circumstances this will only occur when the commitlog is first opened. Although
+		// it can also happen if something goes wrong during the asynchronous reset of the secondary
+		// writer in which case this path will try again, but synchronously this time.
+		primaryWriterFlushFn := l.newOnFlushFn(&l.writerState.primaryWriter)
+		secondaryWriterFlushFn := l.newOnFlushFn(&l.writerState.secondaryWriter)
+		l.writerState.primaryWriter.writer = l.newCommitLogWriterFn(primaryWriterFlushFn, l.opts)
+		l.writerState.secondaryWriter.writer = l.newCommitLogWriterFn(secondaryWriterFlushFn, l.opts)
+
+		primaryFile, err := l.writerState.primaryWriter.writer.Open()
+		if err != nil {
+			return nil, err
+		}
+
+		secondaryFile, err := l.writerState.secondaryWriter.writer.Open()
+		if err != nil {
+			return nil, err
+		}
+
+		l.writerState.activeFiles = persist.CommitLogFiles{primaryFile, secondaryFile}
+		return l.writerState.activeFiles, nil
 	}
 
-	l.writerState.activeFile = &file
+	// Swap the primary and secondary writers so that the secondary becomes primary and vice versa.
+	// This consumes the standby secondary writer, but a new one will be prepared asynchronously by
+	// resetting the formerly primary writer.
+	prevPrimary := l.writerState.primaryWriter
+	l.writerState.primaryWriter = l.writerState.secondaryWriter
+	l.writerState.secondaryWriter = prevPrimary
+	l.startSecondaryWriterAsyncReset()
 
-	return file, nil
+	var (
+		// Determine the persist.CommitLogFile for the not-yet-created secondary file so that the
+		// ActiveLogs() API returns the correct values even before the asynchronous reset completes.
+		primaryFile   = l.writerState.activeFiles[1]
```

Comment:
Does it make slightly more sense to access the "last index" than just 1?  i.e. 

```go
primaryFile = l.writerState.activeFiles[len(l.writerState.activeFiles)-1]
```

I know it's longer but seems more easy to read (strangely) since I know it's always getting the last entry.

Not too fussed though.

Reply:
I prefer the existing one because there can only ever be two and the "last" logic feel more confusing to me because it implies there is may be more than two

Reply:
Sure thing.